src/transformers/models/parakeet/modeling_parakeet.py

```diff
@@ -0,0 +1,602 @@
+        input_shape = hidden_states.shape[:-1]
+        batch_size, seq_length = input_shape
+        hidden_shape = (batch_size, seq_length, -1, self.head_dim)
+
+        query_states = self.q_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        key_states = self.k_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+        value_states = self.v_proj(hidden_states).view(hidden_shape).transpose(1, 2)
+
+        attention_interface: Callable = eager_attention_forward
+        if self.config._attn_implementation != "eager":
+            attention_interface = ALL_ATTENTION_FUNCTIONS[self.config._attn_implementation]
+
+        if position_embeddings is not None:
+            query_states_with_bias_u = query_states + self.bias_u.view(
+                1, self.config.num_attention_heads, 1, self.head_dim
+            )
+            query_states_with_bias_v = query_states + self.bias_v.view(
+                1, self.config.num_attention_heads, 1, self.head_dim
+            )
+
+            relative_key_states = self.relative_k_proj(position_embeddings)
+            relative_key_states = relative_key_states.view(1, -1, self.config.num_attention_heads, self.head_dim)
+
+            # terms (b) and (d)
+            matrix_bd = query_states_with_bias_v @ relative_key_states.permute(0, 2, 3, 1)
+            matrix_bd = self.rel_shift(matrix_bd)
+            matrix_bd = matrix_bd[..., :seq_length]
+            matrix_bd = matrix_bd * self.scaling
+
+            if attention_mask is not None:
+                matrix_bd = matrix_bd.masked_fill_(attention_mask.logical_not(), -10000.0)
+
+        # will compute matrix_ac - terms (a) and (c) - and add matrix_bd
+        attn_output, attn_weights = attention_interface(
+            self,
+            query_states_with_bias_u,
+            key_states,
+            value_states,
+            matrix_bd if position_embeddings is not None else attention_mask,
+            dropout=0.0 if not self.training else self.attention_dropout,
+            scaling=self.scaling,
+            **kwargs,
+        )
+
+        # to match the original implementation
+        # cf https://github.com/NVIDIA/NeMo/blob/620d2ba07835c230b2e1ee25fe1322504ce01d79/nemo/collections/asr/parts/submodules/multi_head_attention.py#L336-L340
+        if attention_mask is not None:
+            all_masked_rows = torch.all(~attention_mask, dim=-1).transpose(1, 2)
+            all_masked_rows = all_masked_rows.unsqueeze_(-1)
+            attn_output = attn_output.masked_fill(all_masked_rows, 0.0)
+
+        attn_output = attn_output.reshape(*input_shape, -1).contiguous()
+        attn_output = self.o_proj(attn_output)
+        return attn_output, attn_weights
+
+    def rel_shift(self, attention_scores):
+        """Relative position shift for Shaw et al. style attention."""
+        batch_size, num_heads, query_length, position_length = attention_scores.shape
+        attention_scores = nn.functional.pad(attention_scores, pad=(1, 0))
```

Comment:
so we pad in the position length on the right, but then we view something that was:
`batch_size, num_heads, query_length, position_length` to `batch_size, num_heads, position_length+, query_length`, and then we ignore the first one (the padding). 
Would be nice to have a small intuitive explanaiton of what is happening here!

Reply:
@eustlb?